cmdstanpy/stanfit.py

```diff
@@ -958,6 +965,61 @@ def draws_pd(
             return self._draws_pd
         return self._draws_pd[mask]
 
+    def draws_xr(
+        self, vars: List[str] = None, inc_warmup: bool = False
+    ) -> "xr.Dataset":
+        """
+        Returns the sampler draws as a xarray Dataset.
+        :param vars: optional list of variable names.
+        :param inc_warmup: When ``True`` and the warmup draws are present in
+            the output, i.e., the sampler was run with ``save_warmup=True``,
+            then the warmup draws are included.  Default value is ``False``.
+        """
+        if not XARRAY_INSTALLED:
+            raise RuntimeError(
+                "xarray is not installed, cannot produce draws array"
+            )
+        if inc_warmup and not self._save_warmup:
+            self._logger.warning(
+                "draws from warmup iterations not available,"
+                ' must run sampler with "save_warmup=True".'
+            )
+
+        if vars is None:
+            vars = self.stan_vars_dims.keys()
+
+        self._assemble_draws()
+
+        num_draws = self.num_draws_sampling
+        meta = self._metadata.cmdstan_config
+        attrs = {
+            "stan_version": f"{meta['stan_version_major']}."
+            f"{meta['stan_version_minor']}.{meta['stan_version_patch']}",
+            "model": meta["model"],
+            "num_unconstrained_params": self.num_unconstrained_params,
+            "num_draws_sampling": num_draws,
+        }
+        if inc_warmup and self._save_warmup:
+            num_draws += self.num_draws_warmup
+            attrs["num_draws_warmup"] = self.num_draws_warmup
+
+        data = {}
+        coordinates = {"chains": self.chain_ids, "draws": np.arange(num_draws)}
+        dims = ("draws", "chains")
+        for var in vars:
+            draw1 = 0
+            if not inc_warmup and self._save_warmup:
+                draw1 = self.num_draws_warmup
+            col_idxs = self._metadata.stan_vars_cols[var]
+            var_dims = dims + tuple(
+                f"{var}_dim_{i}" for i in range(len(self.stan_vars_dims[var]))
+            )
+            data[var] = (var_dims, np.squeeze(self._draws[draw1:, :, col_idxs]))
+
+        return xr.Dataset(data, coords=coordinates, attrs=attrs).transpose(
+            'chains', 'draws', ...
```

Comment:
The first order reason is that it made more sense to me that the underlying dimensionality be with chain first, so that if you selected model.draws_xr().alpha[0] you get all the draws for a single chain, rather than the first draw for each of the four chains.

Secondly, the first thing I did after writing this code was test it by using a couple arviz functions, which all complained if I didn't do a transpose, so I figured it would be a good ideal.

```python
/home/brian/miniconda3/envs/stan/lib/python3.9/site-packages/arviz/data/base.py:169: UserWarning: More chains (1000) than draws (4). Passed array should have shape (chains, draws, *shape)
  warnings.warn(
```

Reply:
Ok, sounds reasonable.

Reply:
(arviz should move to named dimensions only at some points, but until that is reality, going this is ok)

Reply:
if you wanted something other than all the draws for a variable, you would be looking at a chain's worth of draws to check for R-hat or the like - does this change the underlying data layout?  the draws nd array is column major so that you get the right memory locality - https://mc-stan.org/cmdstanpy/api.html#cmdstanpy.CmdStanMCMC.draws

Reply:
For 4 chains and 1000 draws the output is an array of 4 1000-long arrays, rather than the 4000-long the existing draws would be. I believe this would still support the same memory locality